### PR TITLE
(#2674) - avoid unneeded attachment GET requests

### LINKED ISSUE #2674: Use `atts_since` in the replication algorithm
Currently we don't do this, which I guess means we read in every attachment for every revision? Bad for perf.

From the [TouchDB wiki](https://github.com/couchbaselabs/TouchDB-iOS/wiki/Replication-Algorithm):

> Also use the “?atts_since” query parameter to pass a list of revisions that the target already has, so the source can optimize by not including the bodies of attachments already known to the target.

### PR DESCRIPTION
**Pull request by gr2m**
This is Work In Progress, it still needs test, just want to see if it is going in the right direction

ping @nolanlawson @janl

--- 

closes #2674

### CODE REVIEW

**Comment by gr2m** [comment]
I can reproduce the issue locally with `AUTO_COMPACTION=true CLIENT=node npm test` – but can someone explain to me what `AUTO_COMPACTION=true` does exactly, how it changes tests?

**Comment by daleharvey** [comment]
So the tests with `AUTO_COMPACTION` set a global option on PouchDB that means documents will automatically compact on every write, so basically there are never any old revisions, only the current documents + conflicts stored, its essentially the same as calling `compact()` after every write

**Comment by nolanlawson** [comment]
Ah yeah if your test relies on auto-compaction being false, then you need to wrap your test in an if-check to make sure that it's not run for the auto-compaction tests. there are examples of this in `test.compact.js`.

**Comment by gr2m** [comment]
I didn’t write any tests yet. My code changes broke existing tests. I’m a bit confused. I’ll look into it of course but don’t know where to start. If you have any idea / pointers, please let me know

**Comment by gr2m** [comment]
okay that _should_ fix the issue https://github.com/pouchdb/pouchdb/pull/5018/commits/35e10b7d35a9109cb89117fbc979710dfe7c8e89, I’ve missed the skipping of undefined docs here: https://github.com/pouchdb/pouchdb/pull/5018/files#diff-3b1908d46435f961119df0ed4762ec7fL54

**Comment by yaronyg** [comment]
Did anyone figure out how revpos is supposed to work with this? Is revpos obsolete now that we have digest? Or is revpos only meaningful in the context of a particular server instance? In other words if my copy of the PouchDB DB has revpos X for a particular attachment and someone else's copy of the same PouchDB DB has revpos Y for the same version of the same attachment, is that o.k.? I've always been a bit confused by revpos.

**Comment by nolanlawson** [comment]
This looks perfect to me, although I'm kinda surprised it's passing CI, since this contains a semicolon-less style and our eslint config is supposed to require semicolons.

@yaronyg This has nothing to do with revpos. It abandons the concept of revpos entirely. Instead, it just checks to see if there is an attachment available locally with the same digest as the thing we were going to fetch from the remote database, and in that case it skips fetching it.

I will note that this PR will not work for `text/plain` attachments, since those have md5sums generated based on Erlang strings, but that's a minor edge case and the end result is that the replication algorithm will be slightly less efficient for plaintext attachments – not that it will fail entirely. So this PR has a :+1: from me, although I'd like to see the semicolon issue fixed and ideally some more comments to explain the optimization. (Also no worries for a second Travis green checkmark; this has already passed once!)

File: src/replicate/getDocs.js [hunk 1]
```diff
@@ -1,10 +1,49 @@
...
+function fileHasChanged(localDoc, remoteDoc, filename) {
+  return !localDoc._attachments ||
+         !localDoc._attachments[filename] ||
+         localDoc._attachments[filename].digest !== remoteDoc._attachments[filename].digest
+}
+
+function getDocAttachments(db, doc) {
+  var filenames = Object.keys(doc._attachments)
+  return Promise.all(filenames.map(function (filename) {
+    return db.getAttachment(doc._id, filename);
```
Comment (nolanlawson) [inline src/replicate/getDocs.js]: We could also check all non-leaf non-deleted revisions here, but for now I think this is fine. That might be overkill anyway.

**Comment by janl** [comment]
@yaronyg haven’t had time to dig into revpos yet.

Could you give this a spin maybe to see if it helps your use-cases? Thank you! :)

**Comment by yaronyg** [comment]
@janl Sad to say but I'm a few weeks out before I have a solid test case I can use to test this. But given that most of what we do is move around binary attachments anything that stops them from being pulled down when they haven't changed is a huge win!

**Comment by gr2m** [comment]
I added all missing semicolons. Wow, I missed quite a few, that was not intended. 
I’ve also started this PR to make sure that `npm run eslint` will catch missing semicolons in future:
https://github.com/pouchdb/pouchdb/pull/5036

**Comment by gr2m** [comment]
@nolanlawson I suggest we create a follow up issue for this? 

> I will note that this PR will not work for text/plain attachments, since those have md5sums generated based on Erlang strings, but that's a minor edge case and the end result is that the replication algorithm will be slightly less efficient for plaintext attachments – not that it will fail entirely. So this PR has a :+1: from me, although I'd like to see the semicolon issue fixed and ideally some more comments to explain the optimization. (Also no worries for a second Travis green checkmark; this has already passed once!)

**Comment by jo** [comment]
Please note that the attachment digest in Apache CouchDB is [generated from compressed file](https://issues.apache.org/jira/browse/COUCHDB-2251) and therefore depends on the compression configured (Snappy, deflate_N). If we adopt the digest on pull replication (what I think we do) there should be no problem. Also the `text/plain` attachment digest generation should not play a role here in that case. If thats not the case (digests are generated locally on pull replication) this would indeed render this optimisation useless.

**Comment by gr2m** [comment]
I guess I should add tests to it, right? Any pointers on how I could measure the same binary attachment does not requested twice?

**Comment by daleharvey** [comment]
So this is a little hacky, but we have a lot of tests depending on it and if we change how we do it we will just have to change all the tests at the same time, but I think an idea to test this is to overwrite the exposed ajax handler, there is an example @ https://github.com/pouchdb/pouchdb/blob/master/tests/integration/test.replication.js#L4059

This can then count the number of requests made to the attachment and you can assert that there are less requests (the test should fail before, pass after), how does that sound?

**Comment by yaronyg** [comment]
I keep thinking about this PR and about digest. We have visited this issue before, for example, [here](https://github.com/pouchdb/pouchdb/issues/3156). In that bug we ran into a situation that because of how Windows works the digests it generates for its attachments are not the same as the digests generated for the same content on other platforms. The work around in that case was that we added the Windows digest values to our test infrastructure.

But what it means is that this optimization won't work if used against a Windows box or against any CouchDB box where the admin has used the publicly supported admin APIs to change their digest configuration. See [here](http://docs.couchdb.org/en/stable/config/misc.html#attachments) for the options.

This is so fragile it makes me sad.

If we had used revpos instead we wouldn't have these problems. But I recognize that revpos has its own issue that you can end up in a situation where two machines both have the same version of the same attachment but at different rev positions and so there will be a sync even though it isn't needed. The code here would avoid that problem.

But I suppose this code doesn't do any active harm if one accepts the reasonable assumption that even if different servers use different digest approaches with different configurations that the probability of an accidental digest collision is so low that we can ignore it.

So I suppose it's o.k. but it just makes me nervous.

Just thinking out loud in case anyone sees something I missed.

**Comment by jo** [comment]
The question is whether the digest is recalculated on the way in (on http pull replication), which I doubt ot is. Please correct me if I am wrong.

The optimisation introduced in this pull does not depend on how the digest is generated, instead it just compares the ones stored in the docs meta.

Sorry if my previous comment was confusing.

**Comment by yaronyg** [comment]
Sorry, I may be making a bad assumption. So let's expose my assumptions.

Assumptions:
1. The digest recorded locally for a particular attachment was generated locally.
2. The digest retrieved from the remote system via the attachment stub in _changes was generated on the remote system

So when the local logic runs and compares the local digest against the remote digest we are calculating two digests that as shown above can be generated using very different algorithms and settings. As such they may not be reasonably comparable.

That is the crux of my concern.

**Comment by gr2m** [comment]
But aren’t the digests stored in the documents? Or are they generated on the fly? My understanding is that they get generated once and then stored in the documents, and we compare these values, so it doesn’t matter where it was generated in the first place

**Comment by yaronyg** [comment]
No, I believe the hashes are regenerated locally. That is why the tests failed when I started testing on Windows. The remote (Linux) server was generating one ID and the local server was generating its own ID and so things didn't match. If I'm following the code correctly I believe one of the places we do this is [here](https://github.com/pouchdb/pouchdb/blob/9c25a2351c51507e75b5a785a1fcb2c6dad66c33/src/adapters/leveldb/index.js#L667) so we don't just take the remotely generated hash and save it. I believe we regenerate it locally.

BTW, if we were just saving the remote hash that would have some fairly scary security implications especially when replicating filtered DBs where we use filters to prevent send/receiving data from less than fully trusted remote dbs. Since PouchDB de-duplicates attachments via the hash it's critical that we generate the hash ourselves to prevent a remote from gaining access to information they should not have.

Note, btw, that because we use MD5 we are still susceptible to this attack which is why PouchDB really needs to change to a secure hash algorithm like SHA256.

**Comment by mohlsen** [comment]
our team at [@project_stanton](https://twitter.com/project_stanton) had seen https://github.com/pouchdb/pouchdb/issues/2674 that this PR is trying to solve.  We tested this out today, and it solved the issue, but we did run into another bug that may have been caused by this.

We're working on the repro case, and will report here.  It seems to be inconsistent, so we're try to track it down.  It's related to syncing with a remote DB that has attachments that were deleted from docs in later revisions.

**Comment by dakr0** [comment]
As @mohlsen mentioned we tried to test these changes and found some issue with syncing. In our test I used this version of pouchdb with express-pouchdb (that was used as remote server) and also with client app run in browser (that used local db). 
1. So I run client app on two browsers (let’s call them _A_ and _B_). 
2. Then on  client _A_ I created plain document (without any attachments). 
3. After that I put two attachments one by one by using `putAttachment` method. Let’s call them `image1.jpg` and `image2.jpg`. 
4. In next step I synchronized local db on _A_ with remote db (by using `sync` method without any additional options like live mode, retry, etc.). 
5. On client _B_ I synchronized its local db with remote db (in the same way like above).
6. On client _A_ I removed one attachment, for example `image1.jpg` 
7. On _B_ I modified plain content of this document (without touching attachments) and put to local db.
8. Then I synchronized dbs on _A_ as first
9. After that I synchronized dbs on _B_
10. I run client app on another browser (_C_) where local db was empty.
11. Then I started the synchronization process on _C_ and got an error (that `image1.jpg` was not found)
12. As the result the synchronization process failed and the data were not replicated at all (the local db on _C_ was still empty)

**Comment by jo** [comment]
@yaronyg convinced me that attachment digest values are recalculated locally in PouchDB. And I understand the possible open attack vector if they would be taken over.

It would be interesting to know whether Apache CouchDB also recalculates attachment hashes on their way in. This could be easily tested by setting up two CouchDB servers with different compression settings and comparing the digests of a replicated doc.

In general it would be good if CouchDB could provide [reproducible digests, calculated from raw data](https://issues.apache.org/jira/browse/COUCHDB-2252).

**Comment by yaronyg** [comment]
There is nothing to stop PouchDB from using reproducible digests everywhere. CouchDB makes no promises about how the digests are calculated anyway.

**Comment by nolanlawson** [comment]
@yaronyg @jo This discussion has gotten a little off-topic; please follow up on md5sum issues in a separate issue. Yes, the plaintest issue is a problem; you can check the tests in `test.compaction.js` if you don't believe me; there are tests that demonstrate that CouchDB generates different md5sums than PouchDB for plaintext as opposed to non-text attachments ([here's an example](https://github.com/pouchdb/pouchdb/blob/29a0bcb7477cd0e6b88c00feccbca6fa1ee453ba/tests/integration/test.attachments.js#L1240-L1266)).

@gr2m I would be happy to merge this PR even without additional tests, but I'm troubled by folks saying that it caused new issues. @mohlsen @dakr0 if you could find tests to reproduce (as Mocha tests in the test suite preferably) then that would be very helpful. 😃

**Comment by gr2m** [comment]
@dakr0 here is a test case for https://github.com/pouchdb/pouchdb/pull/5018#issuecomment-207550052 by @janl: https://gist.github.com/janl/0c461e810cb3e67e17f2bae61b57c415

it’s fixed via 6715907

I’ll see if I can add tests for the initial case (avoiding unneeded GETs) and Jan’s tests now

**Comment by jo** [comment]
@nolanlawson sorry if I was unclear.

The problem with this approach is that CouchDB calculates digest *after* compression, and compression is configurable. Therefore CouchDB generates different md5sums than PouchDB *most of the time*, not only for plaintext attachments.

**Comment by gr2m** [comment]
@jo so what we do with this PR is an optimisation that works in few cases, but falls back to standard behavior in most, right?

**Comment by gr2m** [comment]
I’ve rebased my branch on latest master and there seems to be an issue with https://github.com/pouchdb/pouchdb/issues/4963 now. I’m looking int it

**Comment by jo** [comment]
@gr2m This works only in the (rare imo) case that [`file_compression`](http://docs.couchdb.org/en/stable/config/couchdb.html#couchdb/file_compression) configuration has been changed to `none` (default is `snappy`). But yes, it falls back to previous behavior otherwise.

I'd argue that this only solves a small part of #2674 and does not justify the added complexity. I'd suggest to implement this optimization based on `revpos` instead, as [outlined in the issue](https://github.com/pouchdb/pouchdb/issues/2674#issuecomment-207604358).

**Comment by janl** [comment]
@jo we've tested this on stock CouchDB and it does work with attachments coming from couch or pouch.

**Comment by dakr0** [comment]
@gr2m I've checked this fix for #5018 and it works. Thanks.

**Comment by nolanlawson** [comment]
Thanks for everyone's input and hard work on this. I will add @janl's test case to the tests and then I think this PR is good to go! @gr2m your `rev` fix seems exactly right; I suspected that would be a good addition as a performance improvement but didn't realize it would also be necessary for correctness. Nice catch.

**Comment by nolanlawson** [comment]
See https://github.com/pouchdb/pouchdb/pull/5067; I think this PR needs a bit more work before we can merge.

**Comment by jo** [comment]
As it turns out [COUCHDB-2252](https://issues.apache.org/jira/browse/COUCHDB-2252) is fixed in current CouchDB version 1.6.1. If anybody could approve it would be great.

This means that my objection is invalid and this digest approach is ok from my side.

:+1:

**Comment by nolanlawson** [comment]
I'm sorry to hold off on this PR; I know everybody is very excited about faster replication! I just want to make sure we've nailed down these errors first. I will look into this the next chance I get unless someone else gets to it sooner.

**Comment by gr2m** [comment]
@nolanlawson I think we nailed it now at https://github.com/pouchdb/pouchdb/pull/5067

**Comment by nolanlawson** [comment]
Rebased on top of the new tests; waiting for a green, and then I'm :+1: https://github.com/pouchdb/pouchdb/pull/5085

**Comment by janl** [comment]
Looks all good now :)

**Comment by nolanlawson** [comment]
Everything looks great! Thanks for the PR, looking forward to faster attachment replication. :)

**Pull request merged**